### PR TITLE
Improvements to CLI errors

### LINKED ISSUE #1465: Error messages for `--limit-storage` etc should use appropriate byte units
The error messages for `--storage-size-limit` print information about the low/high/multiples allowed for the argument. These are printed in bytes (but not labeled as bytes). These arguments accept a variety of units, eg "1024", "1KB" ,"1MB", and it is usually best to provide the argument in such units, not bytes.

Example:

```
$ ./tigerbeetle start --addresses=3000 --limit-storage=0 0_0.tigerbeetle
error: --storage-size-limit: size 0 is below minimum: 1108344832
```

These messages should print their numbers in a more readable unit, probably MB. The `--cache-grid` argument already does so.

This may soon be true for `--memory-lsm-manifest` as well cc https://github.com/tigerbeetle/tigerbeetle/pull/1447

### PR DESCRIPTION
**Pull request by brson**
These changes are primarily concerned with displaying human-readable units for byte-oriented command line arguments.

Firstly, it prints constant limits in whatever unit is appropriate, GB, MB, KB.

Secondly, when it prints the user-supplied value, instead of printing the calculated bytes, it prints the string representation of the value as originally written on the command line.

Instead of

```
$ ./tigerbeetle start --addresses=1001 --limit-storage=1KB 0_0.tigerbeetle
error: --storage-size-limit: size 1024 is below minimum: 1108344832
```

we see

```
$ ./tigerbeetle start --addresses=1001 --limit-storage=1KB 0_0.tigerbeetle
error: --limit-storage: size 1KB is below minimum: 1057MB
```

It also fixes the name of `--limit-storage` in the error messages.

Closes https://github.com/tigerbeetle/tigerbeetle/issues/1465

### CODE REVIEW

File: src/flags.zig
Comment (batiati) [inline src/flags.zig]: I think we can be more precise here, making this type responsible for parsing the user input, so we don't need to store the original string or handle special cases for constants.

For example:

```zig
pub const ByteUnit = enum(u32) {
    bytes = 1,
    kb = 1024,
    mb = .kb * 1024,
    gb = .mb * 1024,
    tb = .gb * 1024,
};

pub const ByteSize = struct {
    value: u64,
    unit: ByteUnit,

    pub fn parse(input: []const u8) ByteSize {
        // Move the "parse_value_size" logic here.
        return .{
            .value = parsed_value,
            .unit = parsed_unit,
        };
    }

    pub fn bytes(size: *const ByteSize) u64 {
        return size.value * @intFromEnum(size.unit);
    }

    pub fn suffix(size: *const ByteSize) []const u8 {
        return switch (size.unit) {
            .bytes => "",
            .kb => "KB",
            .mb => "MB",
            .gb => "GB",
            .tb => "TB",
        };
    }
};

// Printing an error message:
flags.fatal("--limit-storage: size {}{s} is below minimum: {}MB", .{
                    start.limit_storage.value,
                    start.limit_storage.suffix(),
                    @divExact(storage_size_limit_min, 1024 * 1024),
                });
```

**Review by batiati — COMMENTED** [review]
For this PR I think we can focus on only improving the error message.
However, should we support decimal inputs? Like `1.5GB`?

File: src/flags.zig
Comment (brson) [reply src/flags.zig]: I have made this change and pushed it. Some of my changes may not be idiomatic so please give it another look.

e.g. I wasn't able to write a self-referential enum like the one you suggested (`ByteUnit`). I also used an underscore to disambiguate a name.

**Comment by brson** [comment]
> For this PR I think we can focus on only improving the error message. However, should we support decimal inputs? Like `1.5GB`?

Seems like something people will expect to work eventually. Maybe it can wait until we receive requests for it.

File: src/flags.zig
Comment (brson) [inline src/flags.zig]: Is there a prettier way to write this overflow check that discards the actual result?

File: src/flags.zig
Comment (batiati) [inline src/flags.zig]: Since `flag` is only used in the error message, wouldn't it be simpler to remove this parameter and return `!ByteSize` from `parse`?

I think it's ok if the call site handles the error by printing the message with the raw input string.

File: src/flags.zig
Comment (batiati) [inline src/flags.zig]: Let's remove the `default` function, declaring the field `unit: ByteUnit = .bytes`.
It simplifies the initialization e.g. `size: ByteSize = .{ .value = 0 };`

File: src/flags.zig
Comment (batiati) [reply src/flags.zig]: The built-in `@mulWithOverflow` returns the result and the overflow bit. The function `std.math.mul()` is just a wrapper to make it idiomatic with errors;

I think we need to discard the result anyway!

File: src/flags.zig
Comment (batiati) [inline src/flags.zig]: It can be simplified, returning the input value as string instead of the suffix.

```zig
        const unit: struct {
            value_input: []const u8,
            unit: ByteUnit,
        } = unit: inline for (units) |unit| {
            const suffixes = unit[0];
            const unit_kind = unit[1];
            for (suffixes) |suffix_| {
                if (std.mem.endsWith(u8, value, suffix_)) {
                    break :unit .{ .value_input = value[0 .. value.len - suffix_.len], .unit = unit_kind };
                }
            }
        } else break :unit .{ .value_input = value, .unit = ByteUnit.bytes };
```

File: src/flags.zig
Comment (batiati) [reply src/flags.zig]: My bad! TIL that [zig does not allow self-dependency on enums](https://github.com/ziglang/zig/issues/5850), which was surprising since it's not an autonumerated enum.

**Review by batiati — COMMENTED** [review]
Awesome! Just a few comments and we are good to merge!

**Comment by insalt-glitch** [comment]
I know that I have not worked on this code, by no means I intend to stall this, but one thing that I found when looking through the CLI-parsing code: The current parse for byte units does not distinguish between `GB` or `KB` and `GiB` or `KiB`. Usually, `GB=1000^3 bytes` and `GiB=1024^3 bytes` for example, i.e. the difference between metric and binary systems. Maybe this was intended? The current system definitely keeps things simple. However, I think a warning might be appropriate?

File: src/flags.zig
Comment (brson) [reply src/flags.zig]: It might be, but there's a preexisting error handling pattern in this file where every function calls `fatal` directly on errors instead of propagating them; and several parsing functions take this same `flag` parameter as part of that pattern.

If the error handling is not correct here I think it should be addressed for the entire file at once. Can I (should I) file a followup issue about it instead of changing it here?

File: src/flags.zig
Comment (brson) [reply src/flags.zig]: I have made this change.

**Comment by brson** [comment]
> I know that I have not worked on this code, by no means I intend to stall this, but one thing that I found when looking through the CLI-parsing code: The current parse for byte units does not distinguish between `GB` or `KB` and `GiB` or `KiB`. Usually, `GB=1000^3 bytes` and `GiB=1024^3 bytes` for example, i.e. the difference between metric and binary systems. Maybe this was intended? The current system definitely keeps things simple. However, I think a warning might be appropriate?

I was also a little surprised - seeing the support for these units made me think the code would be very pedantic about units, but it is not. I think the current behavior is defensible in that arguably most software people are intending to use base-2 units - I at least am always intending base-2 units even if I write "MB"; but I don't have a strong opinion.

File: src/flags.zig
Comment (brson) [reply src/flags.zig]: I have made this change.

**Comment by brson** [comment]
@batiati I have addressed the review items and this is ready for another look.

File: src/flags.zig
Comment (batiati) [reply src/flags.zig]: I don't think it is necessary. Just that it feels odd when we have encapsulated “parse” as a type function that should know nothing about the error handling. I’d suggest you to keep the error handling function symmetrical to the other ones, but calling the “try ByteSize.parse” that we decoupled. Or, if you find it easier/clearer, you can revert it to a single function outside BiteSize, it is your call. I think the most important is to be able to precisely store and retrieve units either from constants or from the user input.

**Comment by batiati** [comment]
> I know that I have not worked on this code, by no means I intend to stall this, but one thing that I found when looking through the CLI-parsing code: The current parse for byte units does not distinguish between `GB` or `KB` and `GiB` or `KiB`. Usually, `GB=1000^3 bytes` and `GiB=1024^3 bytes` for example, i.e. the difference between metric and binary systems. Maybe this was intended? The current system definitely keeps things simple. 

TigerBeetle aligns everything as power-of-two multiples, so there isn't much practical use for decimal units (sure the user can input 1000000 bytes instead of 1MB, but it will be allocated as the closest power-of-two value anyway).
That said, we have the option of either accepting both MB and MiB as the same thing or being pedantic and not accepting decimal units at all, so for that reason, we choose the best experience over nomenclature.

As my personal opinion, `KB`, `MB`, `GB` and `TB` shouldn't exist as a unit for bytes, and should only be considered as a mistyped version of the correct ones `KiB`, `MiB` , `GiB` and `TiB` 😂

image

**Review by batiati — CHANGES_REQUESTED** [review]
> @batiati I have addressed the review items and this is ready for another look.

Great! Please just settle [the `ByteSize.parse` issue](https://github.com/tigerbeetle/tigerbeetle/pull/1565#discussion_r1503257403) before we merge.

I think we can also rename all occurrences (at least in this file) of inappropriate byte units to the correct form `KiB`, `MiB`, `GiB`, `TiB` (in strings, enums and variable names). But it's your call if you want to do that now or in a follow-up PR.

**Comment by brson** [comment]
@batiati I have made the error-handling changes, reintroducing the free function "parse_value_size" and doing the error reporting there, while "ByteSize.parse" returns an error set.

I've filed two followup issues around unit handling:

- https://github.com/tigerbeetle/tigerbeetle/issues/1672
- https://github.com/tigerbeetle/tigerbeetle/issues/1673

**Comment by brson** [comment]
https://github.com/tigerbeetle/tigerbeetle/pull/1674 will likely conflict with this. I can wait until that lands before rebasing this one.

**Comment by brson** [comment]
@batiati This is rebased over the conflicting PRs and I think all review requests are resolved.

File: src/flags.zig
Comment (batiati) [inline src/flags.zig:374]: As a colateral effect of not accepting MB or GB anymore, this error feels terribly cryptic now (imagine asking for 1gb and seeing this 😬)

Could we include MB/GB/TB only for the purpose of returning an specific error telling the user to use MiB/GiB/TiB instead?

Comment (batiati) [reply src/flags.zig:374]: Or maybe a simpler solution: just replace the error message with something like "invalid unit, it should be MiB, GiB or TiB"

Comment (brson) [reply src/flags.zig:374]: I've addressed this now. I restructured this parser a bit to distinguish between incorrect units and incorrect digits.

File: src/flags.zig
Comment (batiati) [inline src/flags.zig]: It's great to handle  `InvalidUnit` separately!

Just a nit: rename `MissingDigits` with something prefixed with `Invalid*` to align with `InvalidUnit` ... for example, `InvalidSize` or `InvalidAmount`.

Also, I think the old message was more informative for this case `"{s}: expected a size, but found '{s}'"`

File: src/flags.zig
Comment (batiati) [inline src/flags.zig]: Now that we have split the unit into a dedicated `[]const u8` slice, what do you think about using `std.ascii.endsWithIgnoreCase` and eliminating the special cases (i.e. "TiB" and "tib" )? 

I think it's okay if someone inputs "1GIB" or "1Gib" instead of "1GiB."

**Review by batiati — COMMENTED** [review]
Sorry for too much back-and-forth, but I'm dogfooding this change and noticing things that may improve my own experience 😉

**Comment by brson** [comment]
> Sorry for too much back-and-forth, but I'm dogfooding this change and noticing things that may improve my own experience 😉

No worries. I'm happy to do as much rework as needed.

File: src/flags.zig
Comment (brson) [reply src/flags.zig]: Done.

File: src/flags.zig
Comment (brson) [reply src/flags.zig]: Done.

**Comment by brson** [comment]
Ah, don't approve this yet. My last patch introduced a new bug.

**Comment by brson** [comment]
Ok, now it's ready for another review - with the split digits/units the unit comparison needs to be for equality, nod just "ends with".

File: src/flags.zig
Comment (batiati) [inline src/flags.zig]: Shouldn't we preserve the test intent here and use an invalid size for the flag `--size` instead?
We could also add checks to test both `InvalidSize` and `InvalidUnit`.

File: src/flags.zig
Comment (batiati) [inline src/flags.zig]: Even better, we don't need the `units` tuples mapping text to enums anymore! 

```suggestion
        const unit = if (split.unit_input.len > 0)
            unit: inline for (std.enums.values(ByteUnit)) |tag| {
                if (std.ascii.eqlIgnoreCase(split.unit_input, @tagName(tag))) {
                    break :unit tag;
                }
            } else {
                return ByteSizeParseError.InvalidUnit;
            };
        else
            ByteUnit.bytes;
```

File: src/flags.zig
Comment (brson) [reply src/flags.zig]: You are right. I was thinking we needed a test for unusual spellings of "MiB", but there are existing tests that cover that case now, and I lost this case completely.

Comment (batiati) [reply src/flags.zig]: Maybe just a test for the deprecated form "--size=1MB", and for something completely wrong like `--size=INVALID`

Comment (brson) [reply src/flags.zig]: Done.

File: src/flags.zig
Comment (brson) [reply src/flags.zig]: Done.

**Review by batiati — APPROVED** [review]

**Pull request merged**
